master/buildbot/reporters/bitbucketserver.py

```diff
@@ -0,0 +1,168 @@
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+from __future__ import absolute_import
+from __future__ import print_function
+from future.moves.urllib.parse import urlparse
+
+from twisted.internet import defer
+
+from buildbot.process.properties import Interpolate
+from buildbot.process.properties import Properties
+from buildbot.process.results import SUCCESS
+from buildbot.reporters import http
+from buildbot.reporters import notifier
+from buildbot.util import bytes2NativeString
+from buildbot.util import httpclientservice
+from buildbot.util import unicode2bytes
+from buildbot.util.logger import Logger
+
+log = Logger()
+
+# Magic words understood by Bitbucket Server REST API
+INPROGRESS = 'INPROGRESS'
+SUCCESSFUL = 'SUCCESSFUL'
+FAILED = 'FAILED'
+STATUS_API_URL = '/rest/build-status/1.0/commits/{sha}'
+COMMENT_API_URL = '/rest/api/1.0{path}/comments'
+HTTP_PROCESSED = 204
+HTTP_CREATED = 201
+
+
+class BitbucketServerStatusPush(http.HttpStatusPushBase):
+    name = "BitbucketServerStatusPush"
+
+    @defer.inlineCallbacks
+    def reconfigService(self, base_url, user, password, key=None,
+                        statusName=None, startDescription=None,
+                        endDescription=None, verbose=False, **kwargs):
+        yield http.HttpStatusPushBase.reconfigService(
+            self, wantProperties=True, **kwargs)
+        self.key = key or Interpolate('%(prop:buildername)s')
+        self.statusName = statusName
+        self.endDescription = endDescription or 'Build done.'
+        self.startDescription = startDescription or 'Build started.'
+        self.verbose = verbose
+        self._http = yield httpclientservice.HTTPClientService.getService(
+            self.master, base_url, auth=(user, password),
+            debug=self.debug, verify=self.verify)
+
+    @defer.inlineCallbacks
+    def send(self, build):
+        props = Properties.fromDict(build['properties'])
+        results = build['results']
+        if build['complete']:
+            status = SUCCESSFUL if results == SUCCESS else FAILED
+            description = self.endDescription
+        else:
+            status = INPROGRESS
+            description = self.startDescription
+
+        # got_revision could be a string, a dictionary or None
+        got_revision = props.getProperty('got_revision', None)
```

Comment:
maybe I should change the hook to use the source branch commit instead of None (for revision) on pull request events, the only problem I see is that the reporter will be updating the status of a different commit than the one being build (the merge commit)